### ISSUE TITLE
Large data set animation lag

### ISSUE THREAD

**Issue by nandorojo** [coinjar/react-native-wagmi-charts#38]
The line chart starts to drop some frames if the data set gets large on iOS. Under 45 items never drops frames. Under 55 is *pretty* reliable, but occasionally will. Once you get over 70, it's consistently dropping frames on the UI thread when animating. I wonder if there could be optimizations for this. 

Possible ideas:
- Investigate the fastest way to do a for loop, since we're looping through lots of data (`for (let i = 0)` vs `.map`)
- Could we turn the `data` array into a shared value under the hood?
- Since `parse(path)` gets used in a few places, maybe we should offload that to the provider to reduce the number of times it gets calculated.
- Fork the redash helpers and try to optimize them if possible.

These are just a few off-hand ideas. I would have to do more digging.

**Comment by jxom** [comment]
> Could we turn the data array into a shared value under the hood?

Yeah, I think this would be the main catalyst to greatly increasing the performance. I really want to toy with the idea of making a shared value version of the `data` prop. This may create a breaking change though, so we would either have to: document that, or create an alternative prop for shared values (I think I would prefer this method), such as `reanimatedData` or something. 

All the other points are great ideas.

**Comment by nandorojo** [comment]
A different prop could work. It might not be necessary though. Moti lets you pass shared values or the values themselves as props. Under the hood, it checks if it’s a shared value or a normal object inside of a derived value. So we might be able to get away with that.

But you’re right, it would be a breaking change for the consumers using `data` from `useLineChart`. So at the very least, `useLineChart` would need to return a different value to be backwards compatible.

I think a different prop could work well though.

**Comment by jxom** [comment]
Oh right that makes sense! My knee jerk reaction was that converting a regular value to a shared value might jeopardise performance, but thinking about it, it should be okay.

I can put this on my TODOs for this week!

**Comment by nandorojo** [comment]
You might be right tbh, I’m not totally sure. The best possible performance is probably sticking with Reanimated in every step possible. They don’t offer the clearest guidance on performance implications of bringing JS objects into worklets.


Also, not sure if this matters for arrays, but dictionaries going from JS to UI thread don’t preserve key order. I’ve opened an issue already since this was annoying for Moti + transform style props changing order. Since we’re using arrays, this shouldn’t really be an issue if we turn JS values into shared values, but something to keep in mind ahead of time.

**Comment by jxom** [comment]
Been investigating making `data` a shared value under the hood. Seems like this problem is really not trivial as reanimated doesn't really have first-class support for external library imports. We would probably have to find a way to convert libraries such as `d3-scale` and `d3-shape` to be worklet compatible. Will probably need a bit more time to look into this one.

**Comment by nandorojo** [comment]
Relevant: https://github.com/software-mansion/react-native-reanimated/issues/1302

Also: https://github.com/sbycrosz/d3-interpolate-path/tree/feat/react-native-reanimated-support

**Comment by nandorojo** [comment]
@jxom I wonder if we should open this as an issue on Redash and specify the exact functions we want to workletize. I think it would fit better there.

**Comment by nandorojo** [comment]
Also https://twitter.com/terrysahaidak/status/1464311437954854921?s=20

looks like we should do normal for loops. I believe the technical fastest way to do them is `i--`. Curious if it would make a difference here.

**Comment by dchhetri** [comment]
Has there been any progress here?

**Comment by aqigif** [comment]
i wanted to know 
Has there been any progress here?
